feat(jest-mock): add a method to mock a function only for certain arguments

benjaminjkraft opened the pull request:
## Summary

The full motivation is described more fully in #13811, but the basic idea is that a common need is to mock a function only when passed certain arguments. All of the current workarounds are either messier, or require third-party packages that have to implement this in messier ways.

So in this commit I add a new method `mockWhen(args)` that says "mock this function only when passed the given arguments". The arguments can be `expect` matchers, which makes it quite flexible. The caller can then customize the returned mock with `mockImplementation` or any of its siblings as usual. This is all based on a similar function we've been using in tests at my company; but putting it in Jest offers a clearer API since it can be a method alongside all the other `mock*`.

For example (see the tests for more examples):
```ts
const fn = jest.fn(); // or any `MockInstance` from `jest.mock`, `spyOn`, etc.
fn.mockReturnValue('generic output');
fn.mockWhen('special input').mockReturnValue('special output');

expect(fn('arbitrary')).toBe('generic output');
expect(fn('special input')).toBe('special output');
```

Fixes #13811.

## Test plan

Added unit tests.

## TODOs

I didn't add the changelog or documentation yet; I wanted to get confirmation that this is something you're interested in first. I can also add type tests if that's helpful.

mrazauskas commented:
This would be great to add.

Could you look through the discussion in #6180? It can be fine to begin with partial support, of course. Perhaps it isn’t that complicated to cover all mock-return/resolve methods?

benjaminjkraft commented:
> It can be fine to begin with partial support, of course. Perhaps it isn’t that complicated to cover all mock-return/resolve methods?

This API already covers all the methods. (It's equivalent to the one in [this comment](https://github.com/facebook/jest/issues/6180#issuecomment-455015550) from `jest-when`, just a different name. The implementation is much simpler than `jest-when` since we don't need to reimplement anything.) The idea is that `mockWhen` creates and returns a new mock which is used only if the arguments match; you then customize that mock as usual. So in the example I did
```ts
fn.mockWhen('special input').mockReturnValue('special output');
```
but you can just as easily use `mockImplementation`, `mockRejectedValue`, etc.

The only thing I didn't add is a parallel `mockWhenOnce` which would be removed once you call the mock with those arguments. You can do `fn.mockWhen(a).mockReturnValueOnce(b)`, but that has a bit different behavior -- when `fn.mockWhen(a)` is called it removes the return value but not the when. I think that should be possible to add but it would be a lot more invasive since we can't just push onto `specificMockImpls` in the same way -- we'd need our own mapping and logic in `mockConstructor` as well.

cc @timkindberg in case they have any thoughts on how this went in `jest-when`

mrazauskas commented:
Wonderfull! Thanks for clarification.

I think `.whenCalledWith()` reads better than `.mockWhen()`. The rest sounds good to me.

Perhaps for testing you could create new test file exclusively for this API? I though that this isn’t full implementation, just because tests were missing.

timkindberg commented on packages/jest-mock/src/__tests__/index.test.ts:
```diff
@@ -707,6 +707,15 @@ describe('moduleMocker', () => {
+      fn.mockReturnValue('generic output');
+      fn.mockWhen('special input').mockReturnValue('special output');
+
+      expect(fn('arbitrary')).toBe('generic output');
+      expect(fn('special input')).toBe('special output');
```
Maybe add some more calls to this? to be sure it's consistent?

```suggestion
      expect(fn('special input')).toBe('special output');
      expect(fn('arbitrary')).toBe('generic output');
      expect(fn('arbitrary')).toBe('generic output');
      expect(fn('special input')).toBe('special output');
```

timkindberg commented on packages/jest-mock/src/__tests__/index.test.ts:
```diff
@@ -1089,6 +1098,83 @@ describe('moduleMocker', () => {
+
+      expect(fn(3)).toBe('hello');
+      expect(fn('world')).toBe('hello, world');
+      expect(fn('hello', 'friend')).toBe('morning, friend');
```
Maybe add a negative partial matching case.
```suggestion      
      expect(fn('hello', 'friend')).toBe('morning, friend');
      expect(fn('hello', 5)).toBe('hello');
```

timkindberg commented on packages/jest-mock/src/__tests__/index.test.ts:
```diff
@@ -1089,6 +1098,83 @@ describe('moduleMocker', () => {
+      expect(instance.increment('hello')).toBe(-3);
+      expect(instance.increment('goodbye')).toBe(3);
+    });
+  });
```
Very exciting! Thank you for working on this. I hope you don't mind I have some thoughts and questions: 
1. What happens when a mock is reset? Do the when's get cleared?
2. What does the failure console output look like? It does it say something like "Expected to be called with X but was called with Y?"
3. What happens if you call with less than or more than the arguments defined in the mockWhen? I assume it's not a match and undefined is returned?
4. Any ideas on how to support a ["default behavior"](https://github.com/timkindberg/jest-when#supports-default-behavior)? I had to add it to jest-when because ppl were defining mocks in a beforeEach and then wanted to define a "default" behavior in each test... so defaultReturnValue (et al) can be set AFTER any `.whenCalledWith` have been set up and it will still be the "default" behavior. For example with your implementation, could we do `jest.fn().whenCalledWith('foo').mockReturnValue(true).mockReturnValue(false)` and have the second mockReturnValue be the "default"? Or would it override the first mockReturnValue?
5. Any idea how we might support [function matcher](https://github.com/timkindberg/jest-when#supports-function-matchers)? Not necessarily now, but maybe eventually.
6. Any ideas how we might support something like the [`when.allArgs` feature](https://github.com/timkindberg/jest-when#supports-matching-or-asserting-against-all-of-the-arguments-together-using-whenallargs) from jest-when? Again, not now, but eventually. 

Points 5 and 6 are a bit more powerful features but they do unlock the full flexibility that users wanted. Those features came later in jest-when's API though. Especially the allArgs feature, check out some use cases: https://github.com/timkindberg/jest-when/blob/master/src/when.test.js#L268-L339.

benjaminjkraft replied:
Thanks for the thoughts!

> 1. What happens when a mock is reset? Do the when's get cleared?

Yep, they get cleared, just like `mockImplementation`s etc. (Added a test for this.)

> 2. What does the failure console output look like? It does it say something like "Expected to be called with X but was called with Y?"

There's no failure at all: if the when doesn't match then you get whatever other behavior you've configured (or `undefined` if you didn't). If you want to say this function was called only with certain arguments, there are already ways to do that (`expect.toHaveBeenCalledWith`). With this API I want to support the case where you might now (or in the future) call the mock with several sets of arguments, and only care about some of those. (For example, at my job we have `enqueueTask(taskType, ...)`, and a lot of tests want to mock it out, but then do something special for certain task types of interest.) Of course if you want that behavior you can set the default to be to throw.

> 3. What happens if you call with less than or more than the arguments defined in the mockWhen? I assume it's not a match and undefined is returned?

Good question, I had to try this to find out. (And I've added a test case.) It's not a match, unless the difference is a present/absent `undefined`. (So `whenCalledWith(1)` will not match `fn(1, 1)`, but will match `fn(1, undefined)`.) In principle we could require the number of arguments match exactly but I think this is probably the least surprising behavior?

> 4. Any ideas on how to support a ["default behavior"](https://github.com/timkindberg/jest-when#supports-default-behavior)? I had to add it to jest-when because ppl were defining mocks in a beforeEach and then wanted to define a "default" behavior in each test... so defaultReturnValue (et al) can be set AFTER any `.whenCalledWith` have been set up and it will still be the "default" behavior. For example with your implementation, could we do `jest.fn().whenCalledWith('foo').mockReturnValue(true).mockReturnValue(false)` and have the second mockReturnValue be the "default"? Or would it override the first mockReturnValue?

Interesting question. You can as you probably saw already do
```ts
fn.mockReturnValue(defaultValue)
fn.whenCalledWith(1).mockReturnValue(specialValue)
```
But they have to be in that order; if you do the opposite order you just always get the default. We could change that -- say `whenCalledWith` always takes precedence? I feel like that would be a little confusing since `whenCalledWith(expect.any(number))` doesn't take precedence over `whenCalledWith(1)` (and can't, really, in general -- you don't always have a total ordering), but it would support the case you describe.

Of course if callers really want, they can get the behavior you describe with the following, it's just a bit more awkward.
```ts
let defaultValue = defaultDefaultValue
fn.mockReturnValue(defaultValue)
fn.whenCalledWith(1).mockReturnValue(specialValue)
defaultValue = otherDefaultValue
```

> 5. Any idea how we might support [function matcher](https://github.com/timkindberg/jest-when#supports-function-matchers)? Not necessarily now, but maybe eventually.

Maybe add support for `expect.satisfies(myPredicate)` to the [asymmetric matchers](https://jestjs.io/docs/expect#asymmetric-matchers)? I feel like that would be generally useful. (Or of course callers can do that themselves via `expect.extend` if they have a common predicate they want to use.)

> 6. Any ideas how we might support something like the [`when.allArgs` feature](https://github.com/timkindberg/jest-when#supports-matching-or-asserting-against-all-of-the-arguments-together-using-whenallargs) from jest-when? Again, not now, but eventually.

Interesting question. This does seem like a useful thing to support eventually.

I want the answer to be `fn.whenCalledWith(...expect.arrayOf(expect.any(Number)))`. Right now there's no `expect.arrayOf` but that's easy to add (or for users to add) if it proves useful. The bigger problem is that you can't actually spread a matcher. In principle we could do something magical to make it work (i.e. have matchers have an iterator that returns a sentinel value, and then if `equals` sees an array with that value, it "expands" it for the purposes of the comparison); that could get a bit confusing and/or complicated though.

Or if we're ok with a special API, it would be easy enough to later implement `fn.whenCalledWithAll(expect.any(Number))` or `fn.whenCalledWith(jest.allArgs(...))`.

timkindberg replied:
For the default thing (which I anticipate devs asking for at some point like they did with jest-when) maybe we can add a `whenCalledWithDefault` or something. In jest-when I added `defaultReturnValue` and `defaultResolvedValue` etc which is a lot of API surface to add and document. So having just one more `whenCalledWith` function instead is easier. e.g. 

```ts
fn.whenCalledWith(1).mockReturnValue(specialValue)
fn.defaultCalledWith.mockReturnValue(defaultValue)
```

This seems like it might not be so hard to add but I'm not sure.

Thanks for all your other responses, everything looks good.

mrazauskas commented on packages/jest-mock/src/__tests__/whenCalledWith.test.ts:
Tests look very good. Could you add two more, please: for `jest.resetAllMocks()` and `mockFn.withImplementation()`. These have some specific logic. Current test do not cover it.

mrazauskas reviewed (commented):
Looks great for my eye. Documentation and change log entry are missing (;

benjaminjkraft commented:
Finally got around to updating this. I added several more tests, and some type tests, and docs. Thanks for all the comments, I think I've addressed them all now.

I recommend reading the tests relating to `mock*Once` and/or `withImplementation` as they explain the arguably-surprising interaction between the two. My experience in our codebase is the issue doesn't come up a ton -- part of the value of `whenCalledWith` is you can avoid the (often more fragile) `mock*Once` entirely -- but let me know if you think it should be documented explicitly. We could also change the behavior (although it's unclear what behavior would be clearer on the whole), or remove `mock*Once` and `withImplementation` from the `whenCalledWith` output.

mrazauskas commented on docs/MockFunctionAPI.md:
````diff
@@ -273,6 +273,58 @@ mockFn(); // 'default'
+mockFn('jest'); // "Testing is great!"
+```
+
+The arguments may also be [`expect` matchers](ExpectAPI.md#matchers):
````
```suggestion
The arguments may also be [`expect` asymmetric matchers](ExpectAPI.md#asymmetric-matchers):
```

mrazauskas commented on docs/MockFunctionAPI.md:
```diff
@@ -273,6 +273,58 @@ mockFn(); // 'default'
+mockFn.mockReturnValue('Hello, world!');
+mockFn.whenCalledWith('jest').mockReturnValue('Testing is great!');
+
+mockFn('world'); // "Hello, world!"
+mockFn('jest'); // "Testing is great!"
```
```suggestion
mockFn('world'); // 'Hello, world!'
mockFn('jest'); // 'Testing is great!'
```

benjaminjkraft replied:
hehe good catch!

mrazauskas commented on docs/MockFunctionAPI.md:
```diff
@@ -273,6 +273,58 @@ mockFn(); // 'default'
+mockFn.mockReturnValue('Hello, world!');
+mockFn.whenCalledWith('jest').mockReturnValue('Testing is great!');
+
+mockFn('world'); // "Hello, world!"
+mockFn('jest'); // "Testing is great!"
```
```suggestion
mockFn('world'); // 'Hello, world!'
mockFn('jest'); // 'Testing is great!'
```

mrazauskas commented on docs/MockFunctionAPI.md:
```diff
@@ -273,6 +273,58 @@ mockFn(); // 'default'
+  .whenCalledWith(expect.any(String))
+  .mockReturnValue('called with a string');
+
+mockFn(3); // "called"
+mockFn('jest'); // "called with a string"
```
```suggestion
mockFn(3); // 'called'
mockFn('jest'); // 'called with a string'
```

mrazauskas commented on docs/MockFunctionAPI.md:
```diff
@@ -273,6 +273,58 @@ mockFn(); // 'default'
+  .whenCalledWith(expect.any(String))
+  .mockReturnValue('called with a string');
+
+mockFn(3); // "called"
+mockFn('jest'); // "called with a string"
```
```suggestion
mockFn(3); // 'called'
mockFn('jest'); // 'called with a string'
```

mrazauskas reviewed (commented):
Thanks! Looks good for my eye. Just a few minor details to polish in documentation.

timkindberg commented on packages/jest-mock/src/__tests__/whenCalledWith.test.ts:
```diff
@@ -0,0 +1,282 @@
+
+    expect(fn('arbitrary')).toBe('generic output');
+    expect(fn('special input')).toBe('generic output');
+  });
```
Hmm I'm not sure I agree with this behavior. I guess it makes sense it would clobber.

benjaminjkraft replied:
Yeah, I know some libraries do this where a `when` takes precedence over a regular mock. But then you also want `when('special', 3)` to take precedence over `when(Any(), 3)`, and with arbitrary asymmetric matchers that gets impossible very fast. (Not to mention: how does that interact with `mock*Once`/`withImplementation`?) So instead we just give later mocks precedence, such that the behavior is clear.

timkindberg commented on packages/jest-mock/src/index.ts:
```diff
@@ -806,6 +808,28 @@ export class ModuleMocker {
+        f.mockImplementation(combinedImplementation);
+
+        return thisImplementation;
+      };
```
Amazing that this small amount of code is able to add all of those features

benjaminjkraft commented on packages/jest-mock/src/__tests__/whenCalledWith.test.ts:
```diff
@@ -0,0 +1,282 @@
+    expect(fn('special input')).toBe('extra special output');
+    expect(fn('arbitrary')).toBe('generic output');
+    expect(fn('special input')).toBe('special output');
+  });
```
oops, it totally is!

ologbonowiwi commented:
Hey folks, any plan to merge this?
